azure-quantum/azure/quantum/workspace.py

```diff
@@ -260,18 +265,34 @@ def list_jobs(
 
         return result
 
-    def get_targets(self) -> Dict[str, List[str]]:
-        """Returns a dictionary of provider IDs and lists of targets
-        that are available.
-
-        :return: Targets, keyed by provider IDs
-        :rtype: Dict[str, List[str]]
+    def _get_target_status(self, name: str, provider_id: str) -> List[Tuple[str, "TargetStatus"]]:
+        """Get provider ID and status for targets"""
+        return [
+            (provider.id, target)
+            for provider in self._client.providers.get_status()
+            for target in provider.targets
+            if (provider_id is None or provider.id.lower() == provider_id.lower())
+                and (name is None or target.id.lower() == name.lower())
+        ]
+
+    def get_targets(
+        self, 
+        name: str = None, 
+        provider_id: str = None,
+        **kwargs
+    ) -> Union["Target", Iterable["Target"]]:
+        """Returns all available targets for this workspace filtered by name and provider ID.
+        
+        :param name: Optional target name to filter by, defaults to None
+        :type name: str, optional
+        :param provider_id: Optional provider Id to filter by, defaults to None
+        :type provider_id: str, optional
+        :return: Targets
+        :rtype: Iterable[Target]
         """
-        return {
-            provider.id: [
-                target.id for target in provider.targets
-            ] for provider in self._client.providers.get_status()
-        }
+        from azure.quantum.target.target_factory import TargetFactory
+        target_factory = TargetFactory(workspace=self)
+        return target_factory.get_targets(name=name, provider_id=provider_id)

```

Comment:
Can we cache the TargetFactory and use it as a singleto?

Reply:
Great idea!